src/node_crypto.cc

```diff
@@ -4825,6 +4826,429 @@ void Scrypt(const FunctionCallbackInfo<Value>& args) {
+          return false;
+      } else {
+        // Encode SEC1 as DER.
+        if (i2d_ECPrivateKey_bio(bio.get(), ec_key.get()) != 1)
+          return false;
+      }
+    }
+
+    BIOToStringOrBuffer(bio.get(), private_key_encoding_.pem_encoded_, privkey);
+    return true;
+  }
+
+  inline void BIOToStringOrBuffer(BIO* bio, bool pem, Local<Value>* out) const {
+    BUF_MEM* bptr;
+    BIO_get_mem_ptr(bio, &bptr);
+    if (pem) {
+      // PEM is an ASCII format, so we will return it as a string.
+      *out = String::NewFromUtf8(env->isolate(), bptr->data,
+                                 NewStringType::kNormal,
+                                 bptr->length).ToLocalChecked();
+    } else {
+      // DER is binary, return it as a buffer.
+      *out = Buffer::Copy(env, bptr->data, bptr->length).ToLocalChecked();
+    }
+  }
+
+ private:
+  CryptoErrorVector errors_;
+  std::unique_ptr<KeyPairGenerationConfig> config_;
+  PublicKeyEncodingConfig public_key_encoding_;
+  PrivateKeyEncodingConfig private_key_encoding_;
+  EVPKeyPointer pkey_;
+};
+
+void GenerateKeyPair(const FunctionCallbackInfo<Value>& args,
+                     unsigned int n_opts,
+                     std::unique_ptr<KeyPairGenerationConfig> config) {
+  Environment* env = Environment::GetCurrent(args);
+  PublicKeyEncodingConfig public_key_encoding;
+  PrivateKeyEncodingConfig private_key_encoding;
+
+  // Public key encoding: type (int) + pem (bool)
+  CHECK(args[n_opts]->IsInt32());
+  public_key_encoding.type_ = static_cast<PKEncodingType>(
+      args[n_opts]->Int32Value(env->context()).ToChecked());
+  CHECK(args[n_opts + 1]->IsBoolean());
+  public_key_encoding.pem_encoded_ =
+      args[n_opts + 1]->BooleanValue(env->context()).ToChecked();
+
+  // Private key encoding: type (int) + pem (bool) + cipher (optional, string) +
+  //                       passphrase (optional, string)
+  CHECK(args[n_opts + 2]->IsInt32());
+  private_key_encoding.type_ = static_cast<PKEncodingType>(
+      args[n_opts + 2]->Int32Value(env->context()).ToChecked());
+  CHECK(args[n_opts + 3]->IsBoolean());
+  private_key_encoding.pem_encoded_ =
+      args[n_opts + 3]->BooleanValue(env->context()).ToChecked();
+  if (args[n_opts + 4]->IsString()) {
+    String::Utf8Value cipher_name(env->isolate(),
+                                  args[n_opts + 4].As<String>());
+    private_key_encoding.cipher_ = EVP_get_cipherbyname(*cipher_name);
+    if (private_key_encoding.cipher_ == nullptr)
+      return env->ThrowError("Unknown cipher");
+
+    // We need to take ownership of the string here, that's why we are not using
+    // String::Utf8Value.
+    CHECK(args[n_opts + 5]->IsString());
+    Local<String> passphrase = args[n_opts + 5].As<String>();
+    private_key_encoding.passphrase_length_ = passphrase->Utf8Length();
+    private_key_encoding.passphrase_.reset(static_cast<char*>(OPENSSL_malloc(
+        private_key_encoding.passphrase_length_ + 1)));
```

Comment:
This doesn’t seem quite right… `passphrase_` is a `shared_ptr<char>`, so it would be freed using `delete` rather than `delete[]`/`free`/`OPENSSL_free`.

If you want to use a `shared_ptr`, you want `shared_ptr<char[]>`, with a custom deleter if necessary; although maybe `MallocedBuffer<char>` from `util.h` does the job too, like we use elsewhere in this file?